L68-core-callback-api.md

```diff
@@ -0,0 +1,106 @@
+Core callback-based completion queue
+----
+* Author(s): vjpai
+* Approver: markdroth
+* Status: Proposed
+* Implemented in: https://github.com/grpc/grpc/projects/12
+* Last updated: June 1, 2020
+* Discussion at
+
+## Abstract
+
+Provide a completion-queue implementation for gRPC core in which the
+completion of an operation triggers the invocation of a registered
+piece of user code rather than posting a tag on an actual queue. The
+new completion queue will have no explicit retrieval function at the
+surface API level.
+
+## Background
+
+gRPC Core uses a completion queue as a notification mechanism of
+completion of RPC stream operations.  `grpc_call_start_batch` (the
+surface API for initiating RPC operation batches) and certain other
+functions associate an RPC operation
+with a _tag_ (currently-unique ID) when it starts. The tag is of type
+`void*` but is not interpreted by gRPC core. The core API user
+(a gRPC wrapped language implementation) will later see that tag when
+accessing the completion queue via a specified accessor function.
+There have long been 2 completion
+queue types:
+
+* `GRPC_CQ_NEXT` : arbitrary number of operations may be queued to it,
+accessed through `grpc_completion_queue_next` which returns the "next"
+item in the queue (but not necessarily FIFO with respect to wall-clock
+time, initiation order, etc.)
+* `GRPC_CQ_PLUCK` : limited number of operations may be queued to it,
+accessed through `grpc_completion_queue_pluck` which allows waiting
+for a specific tag to arrive at the queue and provides that specific
+tag
+
+The completion queue accessor functions may block until a specified
+deadline, and during that time the user's thread is available for the
+library to use in performing RPC operations or polling the operating
+system for the completion of operations.
+
+### Related proposals
+
+* The [EventManager](https://www.github.com/grpc/proposal/XXX) is
+critical to the operation of the callback completion queue since this
+completion queue type will not provide any accessor function to drive
+the progress of the library
+* The callback completion queue is the implementation mechanism for the [C++ callback API](https://www.github.com/grpc/proposal/pull/180)
+
+## Proposal
+
+gRPC core will now support a new completion queue type called
+`GRPC_CQ_CALLBACK`. Although this is called a completion queue because
+it supports the semantic interface of `grpc_completion_queue`,
+operations in this will never be queued in this structure. Instead, a
+completion will cause a user-specified function to be called.
+
+The tags injected into this completion queue
+are still of type `void*` so as not to affect the API of RPC operation
+initiation functions, but these will be interpreted by the
+library. These tags *must* point to a struct of type
+`grpc_completion_queue_functor` which has two user-provided members:
+
+* `void (*functor_run)(struct grpc_completion_queue_functor*, int)` : a pointer to a
```

Comment:
Did you consider having a `void*` argument/field in `grpc_completion_queue_functor` that can be used to pass arbitrary user state? It looks like the functor gets passed back to itself, so that can be used to track state, but that can be a bit awkward to work with.

Reply:
Core is still basically an implementation detail to support the wrapped languages (including C++). So I would expact that the standard way to do this is to actually do what we do in the C++ API implementation:

* the core surface user defines a class that derives from `grpc_completion_queue_functor`
   - `class MyTag : public grpc_completion_queue_functor {`
* the `functor_run` is just a pointer to a static class method
   -  `MyTag() { functor_run = &MyTag::StaticRun; inlineable = 0; ... }`
   - `static void StaticRun(grpc_completion_queue_functor* tag, int ok) {`
* that static method casts the argument to the actual class and invokes the "real" run function on it
   -    `static_cast<MyTag*>(tag)->ActualRun(ok);`
* the state that you're talking about lives in the class which is derived from the core class rather than being pointed at by the core class.

If you actually wanted to do this directly in C, I guess you would follow the same concept but manually extend the struct by making a bigger struct whose first member is the core surface struct and whose remaining members hold the needed state.

Reply:
Yeah, I was asking primarily from a C perspective. I find `void*` easier to work with in these cases. Looking at something like the C# support library, it's a [C library that wraps Core](https://github.com/grpc/grpc/blob/d075d49a1b45bd15dcb88588286d18c262be5735/src/csharp/ext/grpc_csharp_ext.c) providing slightly higher level abstractions over Core that are easy to invoke from C#.

(C is the de facto universal FFI and all that jazz.)

Reply:
Right, even if someone wants to use this in C, I would strongly recommend that their library always have:

```
struct MyFunctor {
  grpc_completion_queue_functor functor;
  int my_other_field1;
  char* my_other_field2;
  ...
};

MyFunctor functor = {.functor = {.functor_run = &RunFunction, .inlineable = false}, .my_other_field1 = 18, .my_other_field2 = "hello world"};

grpc_call_start_batch(..... , &functor);

void RunFunction(grpc_completion_queue_functor *functor_arg, int* ok) {
  MyFunctor* functor = (MyFunctor*)functor_arg;
  ...
}

```

We're aiming to provide basic mechanism in core without dictating policy.

Reply:
Yeah, that's the awkwardness I was talking about. I guess it just like this bikeshed colored `void*`. 😉

Reply:
Yeah, I appreciate your comments, but I think we're going to keep it without a user-specified `void*` in the mix. We have the Python AIO binding using this struct via composition (https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pxd.pxi#L24) and the C++ API implementation using this struct via inheritance (e.g., https://github.com/grpc/grpc/blob/master/include/grpcpp/impl/codegen/callback_common.h#L136-L137) so I think I'd prefer to leave it alone for now. That said, I suppose I could provide a usage example in this gRFC.

Reply:
Added the usage example.